### PR TITLE
R4R: Refactor asset module

### PR DESCRIPTION
**Pull request by chengwenxi**
Resolves: https://github.com/irisnet/irishub/issues/2091

### CODE REVIEW

**Comment by chengwenxi** [comment]
TODO: 

- [ ] query params
- [ ] mint and transfer owner cli test
- [ ] asset fee precision

**Comment by chengwenxi** [comment]
There is no good way to format tokens with different precision, seems we should refactor `coin_type.go` or use `sdk.RegisterDenom`, skip it for now.

File: modules/asset/client/cli/tx.go [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+	)...)
+	return txCmd
+}
+
+// GetCmdIssueToken implements the issue token command
+func GetCmdIssueToken(queryRoute string, cdc *codec.Codec) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "issue-token",
+		Short: "Issue a new token",
+		Example: fmt.Sprintf("%s asset tx issue-token --family=<family> --source=<source> --decimal=<decimal>"+
```
Comment (zhiqiang-bianjie) [inline modules/asset/client/cli/tx.go]: asset tx  -> tx asset

File: modules/asset/internal/keeper/querier.go [hunk 2]
```diff
@@ -32,37 +31,21 @@ func queryToken(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]byte,
...
-		if !found {
-			return nil, sdk.ErrUnknownRequest(fmt.Sprintf("token %s does not exist", params.TokenId))
-		}
 
-		if token.Source == types.GATEWAY {
-			gateway, _ := keeper.GetGateway(ctx, token.Gateway)
-			token.Owner = gateway.Owner
-		}
+	var found bool
+	token, found = keeper.GetToken(ctx, params.TokenID)
```
Comment (zhiqiang-bianjie) [inline modules/asset/internal/keeper/querier.go]: token, found := keeper.GetToken(ctx, params.TokenID)

File: modules/asset/internal/keeper/querier.go [hunk 3]
```diff
@@ -307,10 +166,21 @@ func queryTokenFees(ctx sdk.Context, req abci.RequestQuery, keeper Keeper) ([]by
 		MintFee:  mintFee,
 	}
 
-	bz, err := codec.MarshalJSONIndent(keeper.cdc, fees)
+	bz, err := codec.MarshalJSONIndent(types.ModuleCdc, fees)
```
Comment (zhiqiang-bianjie) [inline modules/asset/internal/keeper/querier.go]: types.ModuleCdc -> keeper.cdc ?

File: types/asset.go [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+	"regexp"
+	"strings"
+)
+
+const FormatUniABSPrefix = "uni:"
+
+// ConvertIdToTokenKeyId return the store key suffix of a token
+func ConvertIdToTokenKeyId(tokenId string) (key string, err error) {
+	var reToken = `[A-Za-z0-9\.]{3,17}`
+	if !regexp.MustCompile(reToken).MatchString(tokenId) {
```
Comment (zhiqiang-bianjie) [inline types/asset.go]: It's better to precompile first，as follow:
` 
     reToken  := regexp.MustCompile(`[A-Za-z0-9\.]{3,17})
     reToken. MatchString(tokenId)
`

File: types/asset.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+	return key, nil
+}
+
+// ConvertDenomToTokenId return the token id of the given denom
+func ConvertDenomToTokenId(denom string) (tokenId string, err error) {
+	var reDnm = `[A-Za-z0-9\.\-]{3,21}`
+	if !regexp.MustCompile(reDnm).MatchString(denom) {
+		return "", fmt.Errorf("token id convert error: invalid denom")
+	}
```
Comment (zhiqiang-bianjie) [inline types/asset.go]: same as above

**Pull request merged**